vochain/state_snapshot.go

```diff
@@ -0,0 +1,422 @@
+	}
+	s.currentTree++
+	if len(s.header.Trees) == s.currentTree {
+		return io.EOF
+	}
+	s.currentTreeReadEnd = s.treeReadPointer +
+		int(s.header.Trees[s.currentTree].Size)
+	_, err := s.file.Seek(int64(s.treeReadPointer), 0)
+	return err
+}
+
+// Read implements the io.Reader interface. Returns io.EOF error when no
+// more bytes available in the current three.
+func (s *StateSnapshot) Read(b []byte) (int, error) {
+	if s.currentTreeReadEnd == s.treeReadPointer {
+		return 0, io.EOF
+	}
+	// shrink b if its too big an return io.EOF
+	if s.treeReadPointer+len(b) > s.currentTreeReadEnd {
+		b = append([]byte{}, b[:s.currentTreeReadEnd-s.treeReadPointer]...)
+		n, err := s.file.Read(b)
+		s.treeReadPointer += n
+		if err == nil {
+			err = io.EOF
+		}
+		return n, err
+	}
+	n, err := s.file.Read(b)
+	s.treeReadPointer += n
+	return n, err
+}
+
+// ReadAll reads the full content of the current tree and returns its bytes.
+// io.EOF error is returned if the bytes have been already read.
+func (s *StateSnapshot) ReadAll() ([]byte, error) {
+	if s.currentTreeReadEnd == s.treeReadPointer {
+		return nil, io.EOF
+	}
+	b := make([]byte, s.currentTreeReadEnd-s.treeReadPointer)
+	n, err := s.file.Read(b)
+	s.treeReadPointer += n
+	return b, err
+}
+
+// Header returns the header for the snapshot containing the information
+// about all the merkle trees.
+func (s *StateSnapshot) Header() *SnapshotHeader {
+	return &s.header
+}
+
+// TreeHeader returns the header for the current tree.
+func (s *StateSnapshot) TreeHeader() *SnapshotHeaderTree {
+	return &s.header.Trees[s.currentTree]
+}
+
+// Path returns the file path of the snapshot file currently used.
+func (s *StateSnapshot) Path() string {
+	return s.path
+}
+
+// Create starts the creation of a new snapshot as a disk file.
+// This method must be called only once and its operation is oposed to `Open`.
+func (s *StateSnapshot) Create(filePath string) error {
+	var err error
+	s.header.Version = snapshotHeaderVersion
+	s.file, err = os.Create(filePath + ".tmp")
+	s.path = filePath
+	return err
+}
+
+// AddTree adds a new tree to the snapshot. `Create` needs to be called first.
+func (s *StateSnapshot) AddTree(name, parent string, root []byte) {
+	s.lock.Lock() // only 1 tree at time is allowed
+	s.header.Trees = append(s.header.Trees, SnapshotHeaderTree{
+		Name:   name,
+		Parent: parent,
+		Root:   root,
+		Size:   0,
+	})
+}
+
+// EndTree finishes the addition of a tree. This method should be called after `AddTree`.
+func (s *StateSnapshot) EndTree() {
+	s.currentTree++
+	s.lock.Unlock()
+}
+
+// Save builds the snapshot started with `Create` and stores in disk its contents.
+// After calling this method the snapshot is finished.
+// `EndTree` must be called before saving.
+func (s *StateSnapshot) Save() error {
+	s.lock.Lock()
+	defer s.lock.Unlock()
+
+	file, err := os.Create(s.path)
+	defer func() {
+		if err := file.Close(); err != nil {
+			log.Warn(err)
+		}
+	}()
```

Comment:
this can be just a `defer file.Close()` since you already check the error at the end, in the happy path. we do not care about close errors if we stopped early because something went wrong already.